Add Reconnecting

CWXDEV opened the pull request:
still WIP, just retargeting 3.9.0, https://github.com/project-fika/Fika-Plugin/issues/28

Lacyway commented on Fika.Core/Coop/Utils/FikaBackendUtils.cs:
```diff
@@ -32,7 +35,15 @@ public static class FikaBackendUtils
+        public static bool SpawnedPlayersComplete = false;
+        public static List<AirdropLootPacket> OldLootPackets = [];
+        public static List<AirdropPacket> OldAirdropPackets = [];
+        public static List<FikaAirdropBox> OldAirdropBoxes = [];
```
I think all these new fields should be cleaned up at a later stage, possibly just adding a new class and containing them in there. We can then `null` or `Dispose()` that class later and clean up the entire reconnect data. In general I try to avoid as many `static` as possible.

CWXDEV replied:
completely agree, can make that a 2.0 if you'd like

Lacyway commented on Fika.Core/Networking/FikaSerializationExtensions.cs:
```diff
@@ -239,5 +242,152 @@ public static Item GetItem(this NetDataReader reader)
+
+        public static void PutSmokeState(this NetDataWriter writer, Throwable smoke)
+        {
+            var smokeToUse = smoke as SmokeGrenade;
```
Use explicit var

CWXDEV replied:
done

CWXDEV commented on Fika.Core/Coop/Airdrops/FikaAirdropsManager.cs:
```diff
@@ -24,16 +25,18 @@ namespace Coop.Airdrops
         public FikaAirdropParametersModel AirdropParameters { get; set; }
         private ManualLogSource Logger { get; set; }
-        float distanceTravelled = 0;
+        public float DistanceTravelled { get; set; } = 0f;
```
done

CWXDEV commented on Fika.Core/Coop/Airdrops/FikaAirdropsManager.cs:
```diff
@@ -24,16 +25,18 @@ namespace Coop.Airdrops
         public bool ClientLootBuilt = false;
         public static int ContainerCount = 0;
+        public AirdropPacket airdropPacketToSend;
+        public AirdropLootPacket lootPacketToSend;
```
removed as unused

CWXDEV commented on Fika.Core/Coop/Components/CoopHandler.cs:
```diff
@@ -246,19 +247,24 @@ private async Task ReadFromServerCharactersLoop()
             {
+                while (!StartSpawning)
+                {
+                    await Task.Delay(1000);
```
Changed to 250, this was an issue towards the beginning with a race condition, this however might not be needed now. i will remove it and double check things still work as intended

CWXDEV commented on Fika.Core/Coop/GameMode/CoopGame.cs:
```diff
@@ -1163,9 +1198,20 @@ private async Task WaitForPlayers()
+                {
+                    do
+                    {
+                        FikaBackendUtils.ScreenController.ChangeStatus($"Loading all expected Players/Bots, {FikaBackendUtils.ReconnectPacket.Value.PlayerCount}/{Singleton<GameWorld>.Instance.AllPlayersEverExisted.Count()}");
```
updated to use second param

CWXDEV commented on Fika.Core/Coop/Players/ObservedCoopPlayer.cs:
```diff
@@ -645,7 +645,10 @@ public override void OnDead(EDamageType damageType)
                     else
                     {
-                        NotificationManagerClass.DisplayWarningNotification($"Group member '{nickname}' has died");
+                        if (!FikaBackendUtils.IsReconnect)
```
changed to use the SpawnedComplete Field, this way it does

CWXDEV commented on Fika.Core/Coop/Patches/LocalGame/TarkovApplication_LocalGameCreator_Patch.cs:
```diff
@@ -169,5 +196,36 @@ public void HandleLoadComplete(IResult error)
+            if (!FikaBackendUtils.ReconnectPacket.Value.IsAlive)
+            {
+                Logger.LogError($"Player: {profileId} was dead, returning to menu");
+                throw new Exception("You were dead, returning to menu");
```
I have actually solved this with making the player "leave", this was an original solution i had, will remove that now.

I call: L1816 in GameMode.cs
`FikaRequestHandler.RaidLeave(body);`

CWXDEV commented:
alright, so i have removed and tested the "startSpawning" conditionals as they are not needed now,
there is currently one issue with a mapping that has been resolved, but its commented out for now,
moved logic for "clearing" statics to after when its used, solves the need to do it twice.

Up to you if you wanna merge it asap or wait for the SPT mapping fix for modules
